includes/class-wcs-action-scheduler-customer-notifications.php

```diff
@@ -0,0 +1,447 @@
+		if ( $interval <= 2 && 'day' === $period ) {
+			return true;
+		}
+
+		return false;
+	}
+
+	public function get_time_offset( $subscription ) {
+		/**
+		 * Offset between a subscription event and related notification.
+		 *
+		 * @since x.x.x
+		 *
+		 * @param int $time_offset
+		 */
+		return apply_filters( 'woocommerce_subscriptions_customer_notification_time_offset', $this->time_offset, $subscription );
+	}
+
+	public function set_time_offset( $time_offset ) {
+		$this->time_offset = $time_offset;
+	}
+
+	public function __construct() {
+		parent::__construct();
+
+		$setting_option    = get_option(
+			WC_Subscriptions_Admin::$option_prefix . WC_Subscriptions_Email_Notifications::$offset_setting_string,
+			[
+				'number' => 3,
+				'unit'   => 'days',
+			]
+		);
+		$this->time_offset = self::convert_offset_to_seconds( $setting_option );
+
+		add_action( 'woocommerce_before_subscription_object_save', [ $this, 'update_notifications' ], 10, 2 );
+	}
+
+	/**
+	 * Calculate time offset in seconds from the settings array.
+	 *
+	 * @param array $offset Format: [ 'number' => 3, 'unit' => 'days' ]
+	 *
+	 * @return float|int
+	 */
+	protected static function convert_offset_to_seconds( $offset ) {
+		$default_offset = 3 * DAY_IN_SECONDS;
+
+		if ( ! isset( $offset['unit'] ) || ! isset( $offset['number'] ) ) {
+			return $default_offset;
+		}
+
+		switch ( $offset['unit'] ) {
+			case 'days':
+				return ( $offset['number'] * DAY_IN_SECONDS );
+			case 'weeks':
+				return ( $offset['number'] * WEEK_IN_SECONDS );
+			case 'months':
+				return ( $offset['number'] * MONTH_IN_SECONDS );
+			case 'years':
+				return ( $offset['number'] * YEAR_IN_SECONDS );
+			default:
+				return $default_offset;
+		}
+	}
+
+	protected function maybe_schedule_notification( $subscription, $action, $timestamp ) {
+		if ( ! WC_Subscriptions_Email_Notifications::notifications_globally_enabled() ) {
+			return;
+		}
+
+		if (
+			! (
+				$subscription->has_status( 'active' )
+				|| $subscription->has_status( 'pending-cancel' )
+			)
+		) {
+			return;
+		}
+
+		if ( self::is_subscription_period_too_short( $subscription ) ) {
+			return;
+		}
+
+		$action_args = self::get_action_args( $subscription );
+
+		$next_scheduled = as_next_scheduled_action( $action, $action_args, self::$notifications_as_group );
+
+		if ( $timestamp === $next_scheduled ) {
+			return;
+		}
+
+		$this->unschedule_actions( $action, $action_args );
+
+		// Only reschedule if it's in the future
+		if ( $timestamp <= time() ) {
+			return;
+		}
+
+		as_schedule_single_action( $timestamp, $action, $action_args, self::$notifications_as_group );
+	}
+
+	/*
+	 * Subtract time offset from given datetime based on the settings and subscription properties and return resulting timestamp.
+	 *
+	 * @param string $datetime
+	 * @param WC_Subscription $subscription
+	 *
+	 * @return int
+	 */
+	protected function sub_time_offset( $datetime, $subscription ) {
```

Comment:
Nit: Maybe rename this to `get_sub_time_offset()`?

Reply:
I dont think this is getting a time offset though, its subtracting an offset. Based on that it should probably be something like `subtract_time_offset()`m

Reply:
Changed to `subtract_time_offset ` in [53bca5e](https://github.com/Automattic/woocommerce-subscriptions-core/pull/648/commits/53bca5e75bc8111aad5e8538c8a196fb8ab138c7).